crypto/fips1402/src/main/java/org/keycloak/crypto/fips/FIPS1402Provider.java

```diff
@@ -209,4 +218,39 @@ public Signature getSignature(String sigAlgName) throws NoSuchAlgorithmException
         return Signature.getInstance(JavaAlgorithm.getJavaAlgorithm(sigAlgName), BouncyIntegration.PROVIDER);
             
     }
+
+    @Override
+    public SSLSocketFactory wrapFactoryForTruststore(Supplier<String> hostnameSupplier, SSLSocketFactory delegate) {
+
+        // See https://downloads.bouncycastle.org/fips-java/BC-FJA-(D)TLSUserGuide-1.0.9.pdf - Section 3.5.2 (Endpoint identification)
+        return new CustomSSLSocketFactory(delegate) {
+
+            @Override
+            protected Socket configureSocket(Socket s) {
+                if (s instanceof SSLSocket) {
+                    SSLSocket ssl = (SSLSocket)s;
+                    String hostname = hostnameSupplier.get();
+                    SNIHostName sniHostName = getSNIHostName(hostname);
+                    if (sniHostName != null) {
+                        SSLParameters sslParameters = new SSLParameters();
+                        sslParameters.setServerNames(Collections.singletonList(sniHostName));
+                        ssl.setSSLParameters(sslParameters);
+                    }
+                }
+                return s;
```

Comment:
What happens if it's null (for whatever reason, for example hostnameSupplier isn't set properly or something)? Does it fail checks, or just it silently ignore hostname checks?

Reply:
It will fail checks (that is the reason why the tests were failing before this change).

For more details, it will try to verify against the hostname from IP address, rather than hostname. But if the certificate doesn't have the reference to IP address like `127.0.0.1` (instead of hostname like `localhost`) in the subject name or subject alternative name, it will fail.

Some more details why this is needed for BC providers and not for sun providers: https://downloads.bouncycastle.org/fips-java/BC-FJA-(D)TLSUserGuide-1.0.9.pdf section 3.5.2